### PR TITLE
Support for image with no annotations in RetinaNet

### LINKED ISSUE #2800: Add support for image with no annotations in RetinaNet
## 🚀 Feature

RetinaNet currently doesn't support training with images that do not have object annotations.
This has been disabled due to potential issues with distributed training.

We should enable this back.

For more context on what issues might be present, see https://github.com/pytorch/vision/pull/2784#discussion_r503816099

### PR DESCRIPTION
**Pull request by datumbox**
Fixes #2800

Merge after #3031, coauthored with @fmassa.

To ensure that the patch in this branch does not affect the performance of the model I retrained it and compared it with master.

### Branch

<details><summary>retinanet_resnet50_fpn + no annotations support: boxAP=36.4</summary>

```
0: IoU metric: bbox
0:  Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.364
0:  Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.558
0:  Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.384
0:  Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.190
0:  Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.401
0:  Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.492
0:  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.314
0:  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.503
0:  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.542
0:  Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.342
0:  Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.584
0:  Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.690
```
</details>

### Master

<details><summary>retinanet_resnet50_fpn: boxAP=36.4</summary>

```
0: IoU metric: bbox
0:  Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.364
0:  Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.558
0:  Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.381
0:  Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.193
0:  Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.400
0:  Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.489
0:  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.314
0:  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.502
0:  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.541
0:  Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.334
0:  Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.584
0:  Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.692
```
</details>

The above master metrics were taken from the recently merged PR on #2954.

### CODE REVIEW

**Comment by datumbox** [comment]
Rebasing this on the latest master fails an unrelated CI test:
- [Windows GPU](https://app.circleci.com/pipelines/github/pytorch/vision/5107/workflows/2ab857fa-c245-47e1-a669-fd8aa98ceca2/jobs/316840/steps)
- [Linux GPU](https://app.circleci.com/pipelines/github/pytorch/vision/5107/workflows/2ab857fa-c245-47e1-a669-fd8aa98ceca2/jobs/316801/steps)

Let's try rebasing later to see if this was the result of a bug nightly upstream.

**Review by fmassa — APPROVED** [review]
Looks awesome, much simpler and more consistent, thanks a lot Vasilis!

**Comment by datumbox** [comment]
There seems to be some upstream change that has side-effects to this PR. It's unclear how this is possible since the changes don't affect the DeformConv2d operator. Details:

- The commit c5d2868, passed all CI test on  20th of November ([CI logs](https://app.circleci.com/pipelines/github/pytorch/vision/5064/workflows/61315be6-1954-46e3-8e58-13b917bd397e))
- Rebasing on latest master with commit 860669e, failed CI tests on 27th of November ([CI logs](https://app.circleci.com/pipelines/github/pytorch/vision/5124/workflows/515ff56b-b8a8-4d94-99ed-3111c6a62331))
- Undoing the rebase and force pushing passes all the CI test on 27th of November ([CI logs](https://app.circleci.com/pipelines/github/pytorch/vision/5126/workflows/1af292f4-24cc-44fe-99d2-7ce3e781288e))

**Comment by datumbox** [comment]
I did a dummy PR #3054 to remove the test introduced and rebase and it works. It seems that the DeformConv2d test for autocast is flaky and we hit a bad seed. I'm going to try to fix this here by setting the seed to a different value.

File: test/test_models_detection_negative_samples.py [hunk 1]
```diff
@@ -128,6 +128,15 @@ def test_forward_negative_sample_krcnn(self):
         self.assertEqual(loss_dict["loss_rpn_box_reg"], torch.tensor(0.))
         self.assertEqual(loss_dict["loss_keypoint"], torch.tensor(0.))
 
+    def test_forward_negative_sample_retinanet(self):
+        model = torchvision.models.detection.retinanet_resnet50_fpn(
+            num_classes=2, min_size=100, max_size=100)
```
Comment (datumbox) [inline test/test_models_detection_negative_samples.py:133]: The addition of this model initialisation leads to the use of a different random seed on a separate test. See below.

File: test/test_ops.py [hunk 2]
```diff
@@ -655,6 +656,7 @@ def script_func_no_mask(x_, offset_, weight_, bias_, stride_, pad_, dilation_):
 
     @unittest.skipIf(not torch.cuda.is_available(), "CUDA unavailable")
     def test_autocast(self):
+        set_rng_seed(0)
```
Comment (datumbox) [inline test/test_ops.py:659]: The DeformConv2d autocast method is flaky and the different random seed causes it to fail. By fixing it here we avoid breaking master. It would be good to investigate whether the test can be changed to make it less flaky. From what I see it's not a problem of tolerance since we used the standard reduced value.

**Review by datumbox — COMMENTED** [review]
Added a couple of comments to explain the flakiness we observed for future reference.

File: test/test_ops.py (hunk 2 above)
Comment (fmassa) [reply test/test_ops.py:659]: Might be good to put a note in the code or in an issue to track that this is flaky. Can you open an issue tracking this?

Comment (datumbox) [reply test/test_ops.py:659]: Issue #3060.
